### PR TITLE
Add option to disable bootmenu edits

### PR DESCRIPTION
**Pull request by colinmitchell-ggl**
Add option to disable bootmenu edits and updated menu logic
Removed the spawned function in ShowMenuAndLoad in favor of
file.SetDeadline(). Switched from Stdin to TTY to support this.

Previous tests for menu.Choose spawned a xterm shell and had to use manual sleeps prevent races.
The menu class used methods and member variables of xterm so a wrapper interface/class were made to support mocking+injection
This shaved ~20s off the test time

Signed-off-by: Colin Mitchell <colinmitchell@google.com>

### CODE REVIEW

**Comment by colinmitchell-ggl** [comment]
Well, test is failing because `race detected during execution of test`
This makes sense as menu_test is implemented as (write -> sleep -> write..)
Unsure how the other tests in the file get around this, but perhaps it's just because an edit does a few extra backn'forths.
Works locally but this might be why there wasn't already an edit test :)

**Comment by rjoleary** [comment]
> Well, test is failing because `race detected during execution of test`
> This makes sense as menu_test is implemented as (write -> sleep -> write..)
> Unsure how the other tests in the file get around this, but perhaps it's just because an edit does a few extra backn'forths.
> Works locally but this might be why there wasn't already an edit test :)

Why was there a sleep in the test in the first place? That seems kind of hacky.

**Comment by colinmitchell-ggl** [comment]
> Why was there a sleep in the test in the first place? That seems kind of hacky.

Because Go brazenly doesn't support inheritance so mocking is always more difficult :) 
Updated with my attempt at mocking this. A wrapper class around xterm was necessary since we used both functions and some exposed member variables. Was able to add a bunch more tests and still shave some time.

The mock terminal is now stubbing ReadLine calls. However, I'm still seeing detected race conditions...

They are from:
go func() { chosen := make(chan Entry) }
got := <-chosen
check values later.

Am I mistaken for understanding the `got := <-chosen` line is supposed to wait for the func to complete?

**Comment by colinmitchell-ggl** [comment]
PTAL

File: pkg/boot/menu/menu.go [hunk 1]
```diff
@@ -76,153 +75,136 @@ func SetInitialTimeout(timeout time.Duration) {
...
-				term.SetPrompt("Select a boot option to edit:\r\n > ")
+			num, err := parseBootNum(choice, entries)
+			if err != nil {
+				fmt.Fprintln(term, err)
+				fmt.Fprintln(term, "Returning to main menu...")
+				continue
+			}
+			entries[num-1].Edit(func(cmdline string) string {
+				fmt.Fprintf(term, "The current quoted cmdline for option %d is:\r\n > %q\r\n", num, cmdline)
+				fmt.Fprintln(term, ` * Note the cmdline is c-style quoted. Ex: \n => newline, \\ => \`)
```
Comment (10000TB) [inline pkg/boot/menu/menu.go:129]: Curious why this is "single quote" ?

File: pkg/boot/menu/menu_term.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	return &xterm{
+		*term.NewTerminal(f, ""),
+		f,
+		oldState,
+	}
+}
+
+func (t *xterm) Close() {
+	_ = term.Restore(int(t.fileInput.Fd()), t.oldState)
```
Comment (10000TB) [inline pkg/boot/menu/menu_term.go:66]: Curious why we don't propagate error up ？

File: pkg/boot/menu/menu_term.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (t *xterm) SetTimeout(dur time.Duration) error {
+	return t.fileInput.SetDeadline(time.Now().Add(dur))
+}
+
+// Sets the timeout for the file and adds a timeout refresh on user entry
+func (t *xterm) SetEntryCallback(f func()) error {
```
Comment (10000TB) [inline pkg/boot/menu/menu_term.go:74]: why do we have this function signature have an error return ? given it never going to return an error. Future proof ? (not clear how it is going to be needed in future)

File: pkg/boot/menu/menu_term.go (hunk 2 above)
Comment (colinmitchell-ggl) [reply pkg/boot/menu/menu_term.go]: We should. I've just got so used to seeing `defer obj.Close()` ignoring all close errors :) 
Fix squashed.

File: pkg/boot/menu/menu.go (hunk 1 above)
Comment (10000TB) [reply pkg/boot/menu/menu.go:129]: I see

**Review by 10000TB — APPROVED** [review]
/LGTM

**Pull request merged**